Add segment reference count and handles change info in ShardCluster

congqixia opened the pull request:
Resolves #16619
Add reference count for each search/query request
For SegmentChangeInfo
- Wait all segments in OnlineList to be loaded
- Add handoff event into pending list
- Wait all segments in OfflineList is not used (reference count = 0)

/kind enhancement

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

letian-jiang commented on internal/querynode/shard_cluster.go:
```diff
@@ -381,20 +400,166 @@ func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
+	sc.segmentCond.L.Unlock()
+}
+
+// waitSegmentsNotInUse waits untial all provided segments is not in use.
```
typo: until

congqixia replied:
Fixed

letian-jiang commented on internal/querynode/shard_cluster.go:
```diff
@@ -381,20 +400,166 @@ func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
+	sc.mut.Lock()
+	defer sc.mut.Unlock()
+
+	token := sc.nextToken.Add(1)
```
If `nextToken` means the next token of `SegmentChangeInfo` to be used, I think we should return the value of the token before incrementing.

congqixia replied:
Fixed

letian-jiang commented on internal/querynode/shard_cluster.go:
```diff
@@ -381,20 +400,166 @@ func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
 	result := make(map[int64][]int64) // nodeID => segmentIDs
-	sc.mut.RLock()
-	defer sc.mut.RUnlock()
+	sc.mut.Lock()
```
Question: why change to mutex lock if we only read guarded variables?

congqixia replied:
`segmentAllocations` actually change the `inUse` reference count now.

letian-jiang replied:
got it

letian-jiang commented on internal/querynode/shard_cluster.go:
```diff
@@ -381,20 +400,166 @@ func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
+	}
+}
+
+// handoffSegments processes the handoff/load balance segments update procedure.
```
typo: HandoffSegments

congqixia replied:
fixed

congqixia committed:
Add segment reference count and handles change info in ShardCluster

Resolves #16619
Add reference count for each search/query request
For SegmentChangeInfo
- Wait all segments in OnlineList to be loaded
- Add handoff event into pending list
- Wait all segments in OfflineList is not used (reference count = 0)

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

letian-jiang commented:
/lgtm
